java/java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToNestedRecordPattern.java

```diff
@@ -0,0 +1,261 @@
+            TreePath t = ctx.getPath();
+            TypeElement type = null;
+            Map<PatternTree, List<PatternTree>> recordComponentMap = new LinkedHashMap<>();
+            DeconstructionPatternTree recordPattern = (DeconstructionPatternTree) t.getLeaf();
+            recordComponentMap = findNested(recordPattern, recordComponentMap);
+            
+            Set<String> localVars = new HashSet<>(localVarList);
+            for (PatternTree p : recordComponentMap.keySet()) {
+                List<PatternTree> bindTree = new ArrayList<>();
+                BindingPatternTree bTree = (BindingPatternTree) p;
+                VariableTree v = bTree.getVariable();
+                type = (TypeElement) wc.getTrees().getElement(TreePath.getPath(t, v.getType()));
+                if (type == null || type.getRecordComponents().size() == 0) {
+                    continue;
+                }
+                outer:
+                for (RecordComponentElement recordComponent : type.getRecordComponents()) {
+                    String name = recordComponent.getSimpleName().toString();
+                    String returnType = recordComponent.getAccessor().getReturnType().toString();
+                    returnType = returnType.substring(returnType.lastIndexOf(".") + 1);
+                    if (userVars.get(v.getName().toString()) != null) {
+                        for (UserVariables var : userVars.get(v.getName().toString())) {
+                            if (var.getMethodName().equals(name)) {
+                                bindTree.add((BindingPatternTree) wc.getTreeMaker().BindingPattern(wc.getTreeMaker().Variable(wc.getTreeMaker().
+                                        Modifiers(EnumSet.noneOf(Modifier.class)), var.getVariable(), wc.getTreeMaker().Identifier(returnType), null)));
+                                continue outer;
+                            }
+                        }
+                    }
+                    String baseName = name;
+                    int cnt = 1;
+                    while (SourceVersion.isKeyword(name) || localVars.contains(name)) {
+                        name = CodeStyleUtils.addPrefixSuffix(baseName + cnt++, "", "");
+                    }
+                    localVars.add(name);
+                    bindTree.add((BindingPatternTree) wc.getTreeMaker().BindingPattern(wc.getTreeMaker().Variable(wc.getTreeMaker().
+                            Modifiers(EnumSet.noneOf(Modifier.class)), name, wc.getTreeMaker().Identifier(returnType), null)));
+                }
+                recordComponentMap.put(p, bindTree);
+            }
+
+            DeconstructionPatternTree d = (DeconstructionPatternTree) createNestedPattern((PatternTree) t.getLeaf(), wc, recordComponentMap);
+            while (t != null && t.getLeaf().getKind() != Tree.Kind.IF) {
+                t = t.getParentPath();
+            }
+            IfTree it = (IfTree) t.getLeaf();
```

Comment:
Please see the comment about searching for the `if` above. If we would search the enclosing block, this may need to be changed to look for immediately enclosing block for the statements we need to remove.

As a note `WorkingCopy.resolveRewriteTarget` might be useful here - even if a tree is re-written several times, this method will return the last rewrite target.